tests/python/unittest/test_meta_schedule_tune_relay.py

```diff
@@ -323,6 +326,227 @@ def get_output(data, lib):
+
+    tune_tasks = list(
+        filter(
+            lambda task: "dense" in task.task_name,
+            extracted_tasks,
+        )
+    )
+
+    with tempfile.TemporaryDirectory() as work_dir:
+        if do_tune:
+            config = ReplayTraceConfig(
+                num_trials_per_iter=64,
+                num_trials_total=64,
+            )
+            database = tune_extracted_tasks(tune_tasks, target, config, work_dir=work_dir)
+        else:
+            database = JSONDatabase(
+                path_workload=osp.join(work_dir, "database_workload.json"),
+                path_tuning_record=osp.join(work_dir, "database_tuning_record.json"),
+            )
+
+            for task in tune_tasks:
+                mod = Parse._mod(task.dispatched[0])
+                workload = database.commit_workload(mod)
+
+                sch = tvm.tir.Schedule(mod)
+                block = sch.get_block("compute")
+                schedule_rule = sch.get(block).annotations["schedule_rule"]
+
+                if "dense_vnni" in schedule_rule:
+                    schedule_dense(block, M, False, sch)
+
+                tune_rec = TuningRecord(sch.trace, [0.0], workload, tvm.target.Target(target), [])
+
+                database.commit_tuning_record(tune_rec)
+
+    with ApplyHistoryBest(database):
+        with tvm.transform.PassContext(
+            opt_level=3,
+            config={"relay.backend.use_meta_schedule": True},
+        ):
+            """
+            The log should say
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_expand_dims
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast_1
+            meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_nn_batch_matmul
+
+            This means batch matmul and others are scheduled by TE, and dense (the one not warned) is found in the
+            meta schedule tuning database during ApplyHistoryBest
+            """
+            lib = relay.build(relay_mod, target=target, params=params)
+
+    runtime = tvm.contrib.graph_executor.GraphModule(lib["default"](dev))
+
+    runtime.set_input("data", data)
+    runtime.run()
+
+    out = runtime.get_output(0).numpy()
+
+    np.testing.assert_equal(out, ref)
+
+
+@pytest.mark.skip("Requires cascadelake")
+def test_tune_relay_manual_tir_vnni():
+    tir.TensorIntrin.register(VNNI_INTRIN, dot_product_desc, dot_product_intrin)
+
+    manual_tir_common(do_tune=False)
+
+    def schedule_rule_dense_vnni(sch, block):
+        schedule_dense(block, None, True, sch)
+        return [sch]
+
+    register_func("meta_schedule.dense_vnni", schedule_rule_dense_vnni)
+
+    # TODO(masahi): Weird error from tuning with CheckSubtreeCompactDataflow in for_kind.cc turned on
+    # manual_tir_common(do_tune=True)
```

Comment:
Currently this results in the following error:
```
[08:11:02] /home/masa/projects/dev/tvm/src/meta_schedule/task_scheduler/task_scheduler.cc:127: Scheduler picks Task #0: "fused_nn_contrib_dense_pack_add_add"
Traceback (most recent call last):
  File "test_meta_schedule_tune_relay.py", line 559, in <module>
    test_tune_relay_manual_tir_vnni()
  File "test_meta_schedule_tune_relay.py", line 547, in test_tune_relay_manual_tir_vnni
    manual_tir_common(do_tune=True)
  File "test_meta_schedule_tune_relay.py", line 485, in manual_tir_common
    database = tune_extracted_tasks(tune_tasks, target, config, work_dir=work_dir)
  File "/home/masa/projects/dev/tvm/python/tvm/meta_schedule/tune.py", line 716, in tune_extracted_tasks
    task_scheduler.tune()
  File "/home/masa/projects/dev/tvm/python/tvm/meta_schedule/task_scheduler/task_scheduler.py", line 60, in tune
    _ffi_api.TaskSchedulerTune(self)  # type: ignore # pylint: disable=no-member
  File "/home/masa/projects/dev/tvm/python/tvm/_ffi/_ctypes/packed_func.py", line 237, in __call__
    raise get_last_ffi_error()
tvm._ffi.base.TVMError: Traceback (most recent call last):
  5: TVMFuncCall
  4: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<void (tvm::meta_schedule::TaskScheduler)>::AssignTypedLambda<tvm::runtime::Registry::set_body_method<tvm::meta_schedule::TaskScheduler, tvm::meta_schedule::TaskSchedulerNode, void, , void>(void (tvm::meta_schedule::TaskSchedulerNode::*)())::{lambda(tvm::meta_schedule::TaskScheduler)#1}>(tvm::runtime::Registry::set_body_method<tvm::meta_schedule::TaskScheduler, tvm::meta_schedule::TaskSchedulerNode, void, , void>(void (tvm::meta_schedule::TaskSchedulerNode::*)())::{lambda(tvm::meta_schedule::TaskScheduler)#1}, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tvm::runtime::TVMRetValue)
  3: tvm::meta_schedule::TaskSchedulerNode::Tune()
  2: tvm::meta_schedule::ReplayTraceNode::GenerateMeasureCandidates()
  1: tvm::meta_schedule::ReplayTraceNode::State::GenerateMeasureCandidates()
  0: tvm::support::parallel_for_dynamic(int, int, int, std::function<void (int, int)> const&) [clone .cold]
  File "/home/masa/projects/dev/tvm/src/support/parallel_for.cc", line 128
RuntimeError: parallel_for_dynamic error with ScheduleError: (not rendered)
```

If I remove https://github.com/apache/tvm/blob/0ddaaa6a7d1009ea7ca8313a51eb19abb8ae7699/src/tir/schedule/primitive/for_kind.cc#L160 it works.

Reply:
I believe @vinx13 and I got the same problem of getting this check pass. The root cause is that compact dataflow check is too strict for tensorization and therefore we need to disable it in mma case.

Reply:
Similar case here: https://github.com/apache/tvm/pull/10420

Reply:
It's not about tensorization, but parallelization. Now I have more information about this error. If I turn on a detailed log, I got this from `ReplayTraceNode`:

```
 for i0_0_i1_0_fused_fused in T.serial(8192):
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            for i0_1 in T.serial(8):
                for ax0_init in T.vectorized(16):
                    with T.block("compute_init"):
                        i = T.axis.spatial(1024, i0_0_i1_0_fused_fused // 64 * 8 + i0_1)
                        j = T.axis.spatial(1024, i0_0_i1_0_fused_fused % 64 * 16 + ax0_init)
                        T.reads()
                        T.writes(compute[i, j])
                        T.block_attr({"schedule_rule":"meta_schedule.dense_vnni", "workload":["dense_vnni.x86", ["TENSOR", [1024, 1024], "uint8"], ["TENSOR", [64, 256, 16, 4], "int8"], None, "int32"]})
                        compute[i, j] = 0
                for ax1_0 in T.serial(256):
                    # tir.Block#1
                    with T.block("compute_update_o"):
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                        i = T.axis.spatial(1024, i0_0_i1_0_fused_fused // 64 * 8 + i0_1)
                        j_o = T.axis.spatial(64, i0_0_i1_0_fused_fused % 64)
                        k_o = T.axis.reduce(256, ax1_0)
                        T.reads(compute[i, j_o * 16 : j_o * 16 + 16], placeholder[i, k_o * 4 : k_o * 4 + 4], placeholder_1[j_o, k_o, 0 : 16, 0 : 4])
                        T.writes(compute[i, j_o * 16 : j_o * 16 + 16])
                        A = T.match_buffer(placeholder[i, k_o * 4 : k_o * 4 + 4], [4], dtype="uint8", offset_factor=1)
                        B = T.match_buffer(placeholder_1[j_o, k_o, 0 : 16, 0 : 4], [16, 4], dtype="int8", offset_factor=1)
                        C = T.match_buffer(compute[i, j_o * 16 : j_o * 16 + 16], [16], dtype="int32", offset_factor=1)
                        A_u8x4: T.uint8x4 = A[T.ramp(0, 1, 4)]
                        A_i32: T.int32 = T.reinterpret(A_u8x4, dtype="int32")
                        B_i8x64: T.int8x64 = B[0, T.ramp(0, 1, 64)]
                        B_i32x16: T.int32x16 = T.reinterpret(B_i8x64, dtype="int32x16")
                        C[T.ramp(0, 1, 16)] = C[T.ramp(0, 1, 16)] + T.call_llvm_pure_intrin(9785, T.uint32(0), T.broadcast(0, 16), T.broadcast(A_i32, 16), B_i32x16, dtype="int32x16")
                for i1_1 in T.vectorized(16):
                    with T.block("T_add_1"):
                        ax0 = T.axis.spatial(1024, i0_0_i1_0_fused_fused // 64 * 8 + i0_1)
                        ax1 = T.axis.spatial(1024, i0_0_i1_0_fused_fused % 64 * 16 + i1_1)
                        T.reads(compute[ax0, ax1], placeholder_2[0, ax1])
                        T.writes(T_add[ax0, ax1])
                        T_add[ax0, ax1] = compute[ax0, ax1] + placeholder_2[0, ax1] + 1
    
Error message: The queried subtree root tir.For#0 in SRef tree does not have compact dataflow, because its child block tir.Block#1 on SRef tree is neither a local complete block nor a local reduction block.
```

So even though I'm doing `parellel` before `decompose_reduction` to workaround that strict check, `ReplayTrace` is trying to apply `parallel` to a schedule that is clearly already `decompose_reduction`-ed. This is very weird...

Reply:
It is solved, there is the default set of post processors for CPU that does additional `parallel` on top of my manual schedule https://github.com/apache/tvm/blob/7ff5c83f2191cdf5c4a9c5dbc752789c75f5dfa8/python/tvm/meta_schedule/tune.py#L121. Since I'm using manual schedule, I don't want any post procs. Disabling them fixed this issue, thanks @vinx13 @junrushao1994